src/v/cloud_roles/azure_vm_refresh_impl.cc

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_roles/azure_vm_refresh_impl.h"
+
+#include "cloud_roles/request_response_helpers.h"
+#include "json/ostreamwrapper.h"
+#include "json/schema.h"
+
+#include <rapidjson/error/en.h>
+
+namespace cloud_roles {
+
+azure_vm_refresh_impl::azure_vm_refresh_impl(
+  net::unresolved_address address,
+  aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    std::move(address), std::move(region), as, retry_params) {}
+
+std::ostream& azure_vm_refresh_impl::print(std::ostream& os) const {
+    fmt::print(os, "azure_vm_refresh_impl{{address:{}}}", address());
+    return os;
+}
+
+ss::future<api_response> azure_vm_refresh_impl::fetch_credentials() {
+    auto client_id_opt = config::shard_local_cfg()
+                           .cloud_storage_azure_managed_identity_id.value();
+    if (unlikely(!client_id_opt.has_value())) {
+        // This implementation requires client_id from config to be set.
+        // Strictly speaking, IMDS service does not require it if there is only
+        // a system-assigned managed identity or only one user-assigned managed
+        // identity but it would be brittle to not specify it, it could break if
+        // the user added a new one. This verification is also performed in
+        // admin/server.cc::patch_cluster_config
+        vlog(
+          clrl_log.error,
+          "missing cloud_storage_azure_managed_identity_id in config");
+        // return value is not a perfect match but it works
+        co_return api_request_error{
+          .status = boost::beast::http::status::bad_request,
+          .reason = "missing cloud_storage_azure_managed_identity_id",
+          .error_kind = api_request_error_kind::failed_abort};
+    }
+
+    // performs this GET
+    // 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://storage.azure.com/&client_id={$client_id}'
+    // to the vm-local IMDS
+    auto req = http::client::request_header{};
+    req.method(boost::beast::http::verb::get);
+    // TODO fix deps and use boost::url
+    req.target(fmt::format(
+      "/metadata/identity/oauth2/token?"
+      "api-version=2018-02-01"
+      "&resource=https%3A%2F%2Fstorage.azure.com%2F"
+      "&client_id={}",
+      client_id_opt.value()));
+    req.set("Metadata", "true");
```

Comment:
what is `Metadata`?

Reply:
https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=linux#security-and-authentication 
it's part of a mitigation to prevent redirection of requests.
since it's a get request, this header (required) make sure that the request originated from the app and it's not unintended/malicious